src/mscorlib/shared/System/IO/Path.Windows.cs

```diff
@@ -87,6 +87,54 @@ public static string GetFullPath(string path)
             return fullPath;
         }
 
+        public static string GetFullPath(string path, string basePath)
+        {
+            if (path == null)
+                throw new ArgumentException(nameof(path));
+
+            if (basePath == null)
+                throw new ArgumentNullException(nameof(basePath));
+
+            if (!IsPathFullyQualified(basePath))
+                throw new ArgumentException(SR.Arg_BasePathNotFullyQualified);
+
+            if (basePath.Contains('\0') || path.Contains('\0'))
+                throw new ArgumentException(SR.Argument_InvalidPathChars);
+
+            if (IsPathFullyQualified(path))
+                return GetFullPath(path);
+
+            int length = path.Length;
+
+            if (PathInternal.IsDevice(basePath))
+            {
+                // paths starting with \\?\ & \\.\
+                return RemoveRelativeSegments(CombineNoChecks(basePath, path), PathInternal.GetRootLength(basePath));
+            }
+            else if ((length >= 1 && PathInternal.IsDirectorySeparator(path[0])))
+            {
+                // path is current drive rooted i.e. starts with \
+                return RemoveRelativeSegments(CombineNoChecks(GetPathRoot(basePath), path.Substring(1)));
```

Comment:
Only the device paths should go through RemoveRelativeSegments as GetFullPathName doesn't correctly handle them. The rest should go through GetFullPath(string) after you combine them.

Reply:
This case is `\Foo` and `C:\Bar`. You're returning `C:\Foo`.